feat: detect obsolete BQ Storage extra at runtime

plamut opened the pull request:
Closes #629.

This PR adds logic to detect obsolete versions of BQ Storage extra and emit a warning or raise an error where applicable.

Traced all BQ Storage-related code and call paths and I think the changes here cover them all. As a bonus, we get a graceful fallback to `tabledata.list` in most cases (a warning is emitted instead of raising an error).

- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

plamut committed:
feat: detect obsolete BQ Storage extra at runtime

plamut committed:
Cover the changes with unit tests

plamut committed:
Skip BQ Storage version tests if extra missing

plamut commented:
Will fix the coverage error son, one line not hit by the tests.

jimfulton commented on google/cloud/bigquery/client.py:
```diff
@@ -464,6 +466,12 @@ def _create_bqstorage_client(self):
             return None
 
+        try:
+            _verify_bq_storage_version()
```
Why not do this check in `Connection.__init__` if the `bqstorage_client` is not None after calling this?

You could even set it to None if the check fails.

jimfulton replied:
I see this is also called in `table.py`, so you have the check in one place.  Maybe pass the given value here and do the create-if-none check here and return None if the check fails even if a non-None value is passed in.

plamut replied:
Sorry, not following, you mean `Connection.__init__` from `dbapi`? That is not necessarily used here, the client methods can hit this path without using `dbapi`.

plamut replied:
> I see this is also called in `table.py`, so you have the check in one place. Maybe pass the given value here and do the create-if-none check here and return `None` if the check fails even if a non-None value is passed in.

I'll have another look at the convoluted call graph that I drew. :) Might actually work, I'll consider it.

jimfulton commented on google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -280,6 +281,8 @@ def _bqstorage_fetch(self, bqstorage_client):
         # bigquery_storage can indeed be imported here without errors.
         from google.cloud import bigquery_storage
 
+        _verify_bq_storage_version()
```
If we do the check in `Connection.__init__`, why do we need to do it here?  Are we afraid that someone would set that attribute?

Why is it fatal here but not in `Connection.__init__`?

plamut replied:
Somebody could create a `Cursor` instance directly, i.e. without going through `Connection.cursor()`:

```py
conn = dbapi.Connection()
cursor = dbapi.Cursor(conn)
```

Not saying this is necessarily a recommended pattern, but `Cursor` is directly exposed in `google.cloud.bigquery.dbapi` after all, thus has its own check.

jimfulton replied:
> Somebody could create a `Cursor` instance directly, i.e. without going through `Connection.cursor()`:
> 
> ```python
> conn = dbapi.Connection()
> cursor = dbapi.Cursor(conn)
> ```

Sure, but the `Connection` constructor will still do the check.  The cursor will still get the storage client from the connection.

I may be misunderstanding you. :)

plamut replied:
Or maybe it's too late here already. :)
Yeah, you're right, `Cursor` takes the client that's stored on the `Connection` instance, will simplify. And I even think there's still one non-graceful code path in `dbpai` (read; no fallback on the REST client), will double check tomorrow.

jimfulton replied:
> > Why is it fatal here but not in `Connection.__init__`?
> 
> This method specifically tries to fetch data over the BQ Storage API, and if it is given an outdated client, it's the caller's fault somewhere higher up the call chain. The method can't guess how to recover, thus at least throws a more informative exception.
> 
> (otherwise an `AttributeError` would be raised a few lines below at `bigquery_storage.types.DataFormat.ARROW`)

See my straw man I just posted for `_create_bqstorage_client`.  If all `bqstorage_client` assignments go through that, then you can never end up with an invalid client and thus never get the `AttributeError` mentioned above.

From your other comments, I think you can see where I'm going. :)

jimfulton commented on google/cloud/bigquery/_pandas_helpers.py:
```diff
@@ -628,6 +629,8 @@ def _download_table_bqstorage(
     # is available and can be imported.
     from google.cloud import bigquery_storage
 
+    _verify_bq_storage_version()
```
Why is this fatal here, but a warning elsewhere?

plamut replied:
Similar reasoning as below, the helper's job is to download data specifically using the BQ Storage API. If it is called with an obsolete `bqstorage_client`, the actual error most likely is somewhere higher up.

Now, the check _might_ be unnecessary if the higher level code is water proof, but that might not be the case.

jimfulton commented on google/cloud/bigquery/magics/magics.py:
```diff
@@ -777,6 +777,12 @@ def _make_bqstorage_client(use_bqstorage_api, credentials, client_options):
         )
         raise customized_error from err
 
+    try:
+        bigquery._helpers._verify_bq_storage_version()
+    except bigquery.LegacyBigQueryStorageError as exc:
+        warnings.warn(str(exc))
+        return None
+
```
Would it make sense to replace this and the creation logic on line 794 with a call to `_create_bqstorage_client`?

plamut replied:
This magics module already contains a bit too much cruft, thus removing some sounds good. :+1:

jimfulton reviewed (commented):
I think we can be a little DRYer if we reuse `_create_bqstorage_client` more after adding some arguments to it.

plamut commented:
> I think we can be a little DRYer if we reuse _create_bqstorage_client more after adding some arguments to it.

Sounds reasonable, I'll probably give it a shot. :+1:

jimfulton commented:
I'm proposing that `_create_bqstorage_client` becomes something like:

```
    def _create_bqstorage_client(self, bqstorage_client=None, client_info=None, client_options=None):
        """Create a BigQuery Storage API client using this client's credentials.

        If a client cannot be created due to missing dependencies, raise a
        warning and return ``None``.

        Returns:
            Optional[google.cloud.bigquery_storage.BigQueryReadClient]:
                A BigQuery Storage API client.
        """
        try:
            from google.cloud import bigquery_storage
        except ImportError:
            warnings.warn(
                "Cannot create BigQuery Storage client, the dependency "
                "google-cloud-bigquery-storage is not installed."
            )
            return None

        try:
            _verify_bq_storage_version()
        except LegacyBigQueryStorageError as exc:
            warnings.warn(str(exc))
            return None

        if bqstorage_client is None:
            bqstorage_client = bigquery_storage.BigQueryReadClient(
                credentials=self._credentials,
                client_info=client_info,
                client_options=client_options,
                )

        return bqstorage_client
```

and maybe gets renamed :) and that it's called _even_ when the caller is given a client.

If the library is out of date, the caller will end up with `None` even if given a value.

plamut committed:
Rename and improve _create_bqstorage_client()

The method is renamed to _ensure_bqstorage_client() and now performs
a check if BQ Storage dependency is recent enough.

plamut committed:
Remove BQ Storage check from dbapi.Cursor

The check is now performed in dbapi.Connection, which is sufficient.

plamut committed:
Remove BQ Storage check in _pandas_helpers

The methods in higher layers already do the same check before
a BQ Storage client instance is passed to
_pandas_helpers._download_table_bqstorage() helper.

plamut committed:
Simplify BQ Storage client factory in magics

Lean more heavily on client._ensure_bqstorage_client() to de-duplicate
logic.

plamut commented:
I checked the call chains and if I didn't miss anything, the only place left where a bad BQ Storage client could leak down is `dbapi.Connection`, which I sealed. Some of the lower-level checks turned out to be redundant, thus removed them:
- In `dbapi.Cursor._bqstorage_fetch()`
- In `_pandas_helpers._download_table_bqstorage()`

Refactored `Client._make_bqstorage_client` as suggested and de-duplicated logic in `magics` and `dbapi`.

Apart from the client's factory, the other central place where we check if BQ Storage client is compatible is `table._validate_bqstorage()` (and issues a warning if it isn't). The top level `table` methods use this check and set `bqstorage_client` to `None`, if necessary, before sending it down the call chain to lower level utility functions.

jimfulton commented on google/cloud/bigquery/magics/magics.py:
```diff
@@ -762,12 +762,12 @@ def _split_args_line(line):
 
 
-def _make_bqstorage_client(use_bqstorage_api, credentials, client_options):
+def _make_bqstorage_client(client, use_bqstorage_api, client_options):
```
Why the credentials drop?

jimfulton reviewed (approved):
LGTM

plamut committed:
Cover missing code lines with tests
